### PR TITLE
Formalise the concept of a Query against QLDB enabling object serialization of parameters and results

### PR DESCRIPTION
**Pull request by alpian**
*Issue #, if available:*

*Description of changes:*
IQuery is an interface which provides the PartiQL statement to run, the parameters (as ValueHolders) to send
and some mechanism of turning the result documents into .NET objects of some type. Both the synchronous
and asynchronous paths support this new function.

A new library has been added to bridge the gap between the Ion Object Mapper and the QLDB driver, making it
easier for customers to use that serializer with the QLDB library.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Amazon.QLDB.Driver.Serialization/Amazon.QLDB.Driver.Serialization.csproj [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net5.0</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="AWSSDK.QLDBSession" Version="3.5.2" /> 
+    <ProjectReference Include="..\Amazon.QLDB.Driver\Amazon.QLDB.Driver.csproj" />
+    <ProjectReference Include="..\..\..\alpian\ion-object-mapper-dotnet\Amazon.Ion.ObjectMapper\Amazon.Ion.ObjectMapper.csproj" />
```
Comment (byronlin13) [inline Amazon.QLDB.Driver.Serialization/Amazon.QLDB.Driver.Serialization.csproj:10]: Since the object mapper isn't released yet, I think we can use [git submodules](https://gist.github.com/gitaarik/8735255) within this repo and then get the build passing again

File: Amazon.QLDB.Driver.IntegrationTests/StatementExecutionWithSerializationTests.cs [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+            {
+                get; set;
+            }
+        }
+
+        [TestMethod]
+        public void Execute_InsertDocument_UsingObjectSerialization()
+        {
+            // Given.
+            // Create Ion struct to insert.
```
Comment (byronlin13) [inline Amazon.QLDB.Driver.IntegrationTests/StatementExecutionWithSerializationTests.cs:108]: Incorrect comment, this is a C# object.

File: Amazon.QLDB.Driver.Serialization/ObjectSerializer.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
+﻿using System;
```
Comment (byronlin13) [inline Amazon.QLDB.Driver.Serialization/ObjectSerializer.cs:1]: Missing license header

File: Amazon.QLDB.Driver/driver/BaseQldbDriverBuilder.cs [hunk 4]
```diff
@@ -134,6 +135,12 @@ public TBuilder WithRetryLogging()
             return this.BuilderInstance;
         }
 
+        public TBuilder WithSerializer(ISerializer serializer)
```
Comment (byronlin13) [inline Amazon.QLDB.Driver/driver/BaseQldbDriverBuilder.cs:138]: Need to add docstring since this is public API

File: Amazon.QLDB.Driver/driver/Serializer.cs [hunk 5]
```diff
@@ -0,0 +1,7 @@
+using Amazon.QLDBSession.Model;
```
Comment (byronlin13) [inline Amazon.QLDB.Driver/driver/Serializer.cs:1]: Missing license header

File: Amazon.QLDB.Driver/driver/Serializer.cs [hunk 6]
```diff
@@ -0,0 +1,7 @@
...
+
+public interface ISerializer
```
Comment (byronlin13) [inline Amazon.QLDB.Driver/driver/Serializer.cs:3]: Missing docstrings

File: Amazon.QLDB.Driver/transaction/AsyncTransactionExecutor.cs [hunk 7]
```diff
@@ -52,6 +54,19 @@ public Task<IAsyncResult> Execute(string statement)
             return this.transaction.Execute(statement);
         }
 
+        /// <summary>
+        /// Execute the IQuery against QLDB and retrieve the result.
+        /// </summary>
```
Comment (byronlin13) [inline Amazon.QLDB.Driver/transaction/AsyncTransactionExecutor.cs:59]: Incomplete docstrings here and missing docstrings in `Query<T>`

File: Amazon.QLDB.Driver/result/generic/AsyncResult.cs [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+            {
+                this.CancellationToken = cancellationToken;
+                this.query = query;
+            }
+
+            new public T Current
+            {
+                get
+                {
+                    return query.Deserialize(this.currentEnumerator.Current);
```
Comment (byronlin13) [inline Amazon.QLDB.Driver/result/generic/AsyncResult.cs:123]: What if `query.serializer` is null? The default behavior in the driver is to set serializer to null

File: Amazon.QLDB.Driver.Serialization/Document.cs [hunk 9]
```diff
@@ -0,0 +1,7 @@
+namespace Amazon.QLDB.Driver.Serialization
+{
+    public class Document
```
Comment (byronlin13) [inline Amazon.QLDB.Driver.Serialization/Document.cs:3]: Where does this get used?

File: Amazon.QLDB.Driver/transaction/AsyncTransactionExecutor.cs [hunk 10]
```diff
@@ -52,6 +54,19 @@ public Task<IAsyncResult> Execute(string statement)
...
+        public  Task<Amazon.QLDB.Driver.Generic.IAsyncResult<T>> Execute<T>(IQuery<T> query)
+        {
+            return this.transaction.Execute(query);
+        }
+
+        public IQuery<T> Query<T>(string statement, params object[] parameters)
```
Comment (kylepbit) [inline Amazon.QLDB.Driver/transaction/AsyncTransactionExecutor.cs:65]: Given that you have to type the Query method, and that Execute is not the same as Execute<T>, why do we need the Query method at all? Why could we not use the presence (or absence) of the type as indication of using serialization, and do away with the QueryData object as an explicit reference. Seems much simpler.

Comment (alpian) [reply Amazon.QLDB.Driver/transaction/AsyncTransactionExecutor.cs:65]: `tx.Query` allows the driver to get access to the desired de/serializer. It is not *required* to create an `IQuery` if you the developer want to create your `IQuery` some other way. I deliberately introduced `IQuery`so thay we could eventually support different types of queries instead of adding more and more methods to the `TransactionExecutor` interface.

File: Amazon.QLDB.Driver.Serialization/Document.cs (hunk 9 above)
Comment (alpian) [reply Amazon.QLDB.Driver.Serialization/Document.cs:3]: Anyone inserting/updating rows in QLDB. i.e. not in the serializer or driver code, but consuming libraries.

File: Amazon.QLDB.Driver.Serialization/Amazon.QLDB.Driver.Serialization.csproj (hunk 1 above)
Comment (alpian) [reply Amazon.QLDB.Driver.Serialization/Amazon.QLDB.Driver.Serialization.csproj:10]: This can't be merged until the object mapper is a public thing, you are correct.

File: Amazon.QLDB.Driver/transaction/AsyncTransactionExecutor.cs (hunk 10 above)
Comment (kylepbit) [reply Amazon.QLDB.Driver/transaction/AsyncTransactionExecutor.cs:65]: For example, adding the following method:

`public Task<Amazon.QLDB.Driver.Generic.IAsyncResult<T>> Execute<T>(string statement, params object[] parameters) {
  return this.transaction.Execute(new QueryData<T>(statement, parameters, serializer);
}
`

File: Amazon.QLDB.Driver/driver/BaseQldbDriverBuilder.cs (hunk 4 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver/driver/BaseQldbDriverBuilder.cs:145]: Added.

File: Amazon.QLDB.Driver/transaction/AsyncTransactionExecutor.cs (hunk 7 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver/transaction/AsyncTransactionExecutor.cs:59]: Added.

File: Amazon.QLDB.Driver.IntegrationTests/StatementExecutionWithSerializationTests.cs (hunk 2 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver.IntegrationTests/StatementExecutionWithSerializationTests.cs]: Corrected.

File: Amazon.QLDB.Driver.Serialization/ObjectSerializer.cs (hunk 3 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver.Serialization/ObjectSerializer.cs]: Added.

File: Amazon.QLDB.Driver/driver/Serializer.cs (hunk 5 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver/driver/Serializer.cs]: Added.

File: Amazon.QLDB.Driver/driver/Serializer.cs (hunk 6 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver/driver/Serializer.cs]: Added.

File: Amazon.QLDB.Driver/result/generic/AsyncResult.cs (hunk 8 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver/result/generic/AsyncResult.cs:123]: Added a null check in [QueryData constructor.](https://github.com/awslabs/amazon-qldb-driver-dotnet/blob/object-serialization-idda-jun-4-2021/Amazon.QLDB.Driver/query/Query.cs#L52-L55)

File: .gitmodules [hunk 11]
```diff
@@ -0,0 +1,3 @@
+[submodule "ion-object-mapper-dotnet"]
```
Comment (byronlin13) [inline .gitmodules:1]: This is temporary to get the build passing. Will be removed once Ion object mapper repos is on NuGet.

File: Amazon.QLDB.Driver.IntegrationTests/Amazon.QLDB.Driver.IntegrationTests.csproj [hunk 12]
```diff
@@ -1,7 +1,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>netcoreapp2.1;netcoreapp3.1;net48</TargetFrameworks>
```
Comment (byronlin13) [inline Amazon.QLDB.Driver.IntegrationTests/Amazon.QLDB.Driver.IntegrationTests.csproj:4]: The object mapper & serialization libraries with target framework: NET 5.0 makes it incompatible with netcoreapp2.1, netcoreapp3.1, net48. 

Need to discuss if we want to force users to update their frameworks to use the object mapper.

File: Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs [hunk 13]
```diff
@@ -106,6 +107,7 @@ public void RunCreateLedger()
 
             return builder.WithQLDBSessionConfig(amazonQldbSessionConfig)
                 .WithLedger(finalLedgerName)
+                .WithSerializer(new ObjectSerializer())
```
Comment (kylepbit) [inline Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs:110]: Is this a requirement, or will it default to this serializer?

File: Amazon.QLDB.Driver.Serialization/ObjectSerializer.cs [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+ */
+
+namespace Amazon.QLDB.Driver.Serialization
+{
+    using System;
+    using System.IO;
+    using Amazon.Ion.ObjectMapper;
+    using Amazon.QLDBSession.Model;
+
+    public class ObjectSerializer : ISerializer
```
Comment (kylepbit) [inline Amazon.QLDB.Driver.Serialization/ObjectSerializer.cs:21]: Given this is public, we should doc it.

File: Amazon.QLDB.Driver/driver/ISerializer.cs [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    public interface ISerializer
+    {
+        /// <summary>
+        /// Serialize a C# object into a ValueHolder object containing the Ion binary or text value.
+        /// </summary>
+        ///
+        /// <param name="o">The C# object to be serialized into ValueHolder.</param>
+        ///
+        /// <returns>The ValueHolder object containing the Ion binary or text value.</returns>
+        ValueHolder Serialize(object o);
```
Comment (kylepbit) [inline Amazon.QLDB.Driver/driver/ISerializer.cs:30]: Any reason this is type object instead of type T?

File: Amazon.QLDB.Driver/query/Query.cs [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+            }
+
+            this.serializer = serializer;
+        }
+
+        public string Statement { get; }
+
+        public ValueHolder[] Parameters
+        {
+            get {
```
Comment (kylepbit) [inline Amazon.QLDB.Driver/query/Query.cs:64]: Nit: `{` on next line

File: Amazon.QLDB.Driver/transaction/TransactionExecutor.cs [hunk 17]
```diff
@@ -66,6 +70,19 @@ public IResult Execute(string statement)
...
+        
+        /// <summary>
+        /// Execute the IQuery against QLDB and retrieve the result.
+        /// </summary>
+        public Amazon.QLDB.Driver.Generic.IResult<T> Execute<T>(IQuery<T> query)
+        {
+            return this.transaction.Execute(query);
+        }
+
+        public IQuery<T> Query<T>(string statement, params object[] parameters)
```
Comment (kylepbit) [inline Amazon.QLDB.Driver/transaction/TransactionExecutor.cs:82]: Need docs here - if this method is being kept and the suggestion for async isn't used.

File: Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs (hunk 13 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs:110]: This is a requirement if the user is planning on using `IQuery`. If the user doesn't pass a serializer, there is a [null check ](https://github.com/awslabs/amazon-qldb-driver-dotnet/blob/object-serialization-idda-jun-4-2021/Amazon.QLDB.Driver/query/Query.cs#L52-L55)in `Query` constructor.

There is an integration [test](https://github.com/awslabs/amazon-qldb-driver-dotnet/blob/object-serialization-idda-jun-4-2021/Amazon.QLDB.Driver.IntegrationTests/StatementExecutionTests.cs#L292-L312) that tests using serializer so it is required here.

File: Amazon.QLDB.Driver.Serialization/ObjectSerializer.cs (hunk 14 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver.Serialization/ObjectSerializer.cs:24]: Added docstrings

File: Amazon.QLDB.Driver/query/Query.cs (hunk 16 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver/query/Query.cs]: Addressed.

File: Amazon.QLDB.Driver/transaction/TransactionExecutor.cs (hunk 17 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver/transaction/TransactionExecutor.cs:96]: Added docstrings

File: Amazon.QLDB.Driver/driver/ISerializer.cs (hunk 15 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver/driver/ISerializer.cs:30]: I think we should update it, I see in the [Object mapper](https://github.com/amzn/ion-object-mapper-dotnet/blob/main/Amazon.Ion.ObjectMapper/IonSerializer.cs#L170-L176) we are using type `T` as well.

Thoughts @alpian  ?

File: Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs (hunk 13 above)
Comment (kylepbit) [reply Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs:110]: Right - by why can't we default to `ObjectSerializer` if one isn't set rather than forcing someone to set one?

Comment (byronlin13) [reply Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs:110]: Theres two issues:

* There is a cyclic dependency error
  * `Amazon.QLDB.Driver.Serialization` is using `Amazon.QLDB.Driver`.
  * If `Amazon.QLDB.Driver` uses `Amazon.QLDB.Driver.Serialization` then it cause the error.

* `Amazon.QLDB.Driver` is using target framework: `netstandard 2.0` and it is incompatible with `Amazon.QLDB.Driver.Serialization`/`Amazon.IonObjectMapper` target framework: `net50`

Comment (kylepbit) [reply Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs:110]: Ah - so we have the interface but not the implementation. Noted - too bad this can't be implicitly loaded if the serialization library is in the runtime (or can it?).

File: Amazon.QLDB.Driver.IntegrationTests/Amazon.QLDB.Driver.IntegrationTests.csproj (hunk 12 above)
Comment (alpian) [reply Amazon.QLDB.Driver.IntegrationTests/Amazon.QLDB.Driver.IntegrationTests.csproj:4]: We can downgrade if we need to. Let's assume we don't need to.

File: Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs [hunk 18]
```diff
@@ -83,6 +85,7 @@ public void RunCreateLedger()
 
         public QldbDriver CreateDriver(
             AmazonQLDBSessionConfig amazonQldbSessionConfig,
+            ISerializer serializer,
```
Comment (alpian) [inline Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs:88]: Why not just default this to `null` rather than changing all the callers?

File: Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs [hunk 19]
```diff
@@ -429,4 +436,57 @@ public string GetDisplayName(MethodInfo methodInfo, object[] data)
             return null;
         }
     }
+
+    class ParameterObject
+    {
+        public string Name
```
Comment (alpian) [inline Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs:442]: We agreed this should not be public

File: Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs [hunk 20]
```diff
@@ -429,4 +436,57 @@ public string GetDisplayName(MethodInfo methodInfo, object[] data)
...
+
+        public override string ToString()
+        {
+            return "<ParameterObject>{ Name: " + Name + " }";
+        }
+    }
+
+    class ResultObject
+    {
+        public string DocumentId
```
Comment (alpian) [inline Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs:458]: Should not be public

File: Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs [hunk 21]
```diff
@@ -429,4 +436,57 @@ public string GetDisplayName(MethodInfo methodInfo, object[] data)
...
+    class MySerialization : ISerializer
+    {
+        private static readonly IValueFactory ValueFactory = new ValueFactory();
+
+        public T Deserialize<T>(QLDBSession.Model.ValueHolder v)
+        {
+            return (T)(object)new ResultObject { DocumentId = "Deserialized using custom serializer" };
+        }
+
+        public QLDBSession.Model.ValueHolder Serialize(object o)
```
Comment (alpian) [inline Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs:478]: Why are you doing this? Just use the Object Mapper! This is what it is for! :D

File: Amazon.QLDB.Driver/client/Session.cs [hunk 22]
```diff
@@ -272,11 +273,30 @@ internal virtual CommitTransactionResult CommitTransaction(string txnId, MemoryS
                     return valueHolder;
                 });
 
+                return await this.ExecuteStatementAsync(txnId, statement, valueHolders.ToArray(), cancellationToken);
+            }
+            finally
```
Comment (simonz-bq) [inline Amazon.QLDB.Driver/client/Session.cs:278]: Why was this moved? This ExecuteStatementAsync seems to call the other, so wouldn't it make more sense for the finally check to be in the other?

File: Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs [hunk 23]
```diff
@@ -45,14 +46,17 @@ public class AsyncQldbDriver : IAsyncQldbDriver
...
         /// <param name="maxConcurrentTransactions">The maximum number of concurrent transactions.</param>
         /// <param name="logger">The logger to use.</param>
+        /// <param name="serializer">The serializer to serialize and deserialize Ion data.</param>
         internal AsyncQldbDriver(
             string ledgerName,
             IAmazonQLDBSession sessionClient,
             int maxConcurrentTransactions,
-            ILogger logger)
+            ILogger logger,
+            ISerializer serializer = null)
```
Comment (simonz-bq) [inline Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs:55]: This seems very odd.

We have a builder to set defaults... so why not set the default in the builder?

File: Amazon.QLDB.Driver/client/Session.cs [hunk 24]
```diff
@@ -319,6 +329,13 @@ internal virtual CommitTransactionResult CommitTransaction(string txnId, MemoryS
                 .GetResult();
         }
 
+        internal virtual ExecuteStatementResult ExecuteStatement(
```
Comment (alpian) [inline Amazon.QLDB.Driver/client/Session.cs:332]: Why was this added?

File: .gitmodules [hunk 25]
```diff
@@ -0,0 +1,3 @@
...
+	path = ion-object-mapper-dotnet
+	url = git@github.com:amzn/ion-object-mapper-dotnet.git
```
Comment (alpian) [inline .gitmodules:3]: Think we should hold off on doing this until we get onto nuget.

**Review by simonz-bq — CHANGES_REQUESTED** [review]

File: Amazon.QLDB.Driver.Serialization/ObjectSerializer.cs [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+namespace Amazon.QLDB.Driver.Serialization
+{
+    using System.IO;
+    using Amazon.Ion.ObjectMapper;
```
Comment (justing-bq) [inline Amazon.QLDB.Driver.Serialization/ObjectSerializer.cs:17]: Update to `IonObjectMapper`

File: Amazon.QLDB.Driver.Serialization/ObjectSerializer.cs [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+        /// Serialize a C# object into a ValueHolder object containing the Ion binary.
+        /// </summary>
+        ///
+        /// <param name="o">The C# object to be serialized into ValueHolder.</param>
+        ///
+        /// <returns>The ValueHolder object containing the Ion binary.</returns>
+        public ValueHolder Serialize(object o)
+        {
+            MemoryStream memoryStream = new MemoryStream();
+            serializer.Serialize(o).CopyTo(memoryStream);
```
Comment (justing-bq) [inline Amazon.QLDB.Driver.Serialization/ObjectSerializer.cs:59]: Any particular reason why we are copying to a new `MemoryStream` rather than just using the stream returned from `serializer.Serialize(o)`?

File: Amazon.QLDB.Driver.Serialization/ObjectSerializer.cs (hunk 26 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver.Serialization/ObjectSerializer.cs]: Done

File: Amazon.QLDB.Driver.Serialization/ObjectSerializer.cs (hunk 27 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver.Serialization/ObjectSerializer.cs:59]: The stream returned by `serializer.Serialize(o)` returns a `Stream` abstract class. It is not of type `MemoryClass`.

I see `Serialize()` has an overloaded function to take in a stream object so I am using that instead.

File: Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs (hunk 19 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs:442]: It is here so we can get the test running. Will update once the object mapper [PR](https://github.com/amzn/ion-object-mapper-dotnet/pull/36) to handle internal/protected internal properties is merged in.

File: Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs (hunk 23 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver/driver/AsyncQldbDriver.cs:55]: The [builder](https://github.com/awslabs/amazon-qldb-driver-dotnet/blob/object-serialization-idda-jun-4-2021/Amazon.QLDB.Driver/driver/BaseQldbDriverBuilder.cs#L29) is setting serializer to `default`.

Removing this because it is redundant

File: Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs (hunk 18 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs:88]: Updated to `default` and reverted changing the callers

File: Amazon.QLDB.Driver/client/Session.cs (hunk 24 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver/client/Session.cs:332]: This is an overloaded function to handle `ValueHolder` objects returned by [IQuery.Parameters](https://github.com/awslabs/amazon-qldb-driver-dotnet/blob/object-serialization-idda-jun-4-2021/Amazon.QLDB.Driver/query/Query.cs#L34) property.

File: Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs (hunk 21 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs:477]: I believe this was added to test the driver using a custom serializer. 

Although I don't see much value in this test as we already have a test to insert/query c# object using `ObjectSerializer` so I will remove this

File: .gitmodules (hunk 25 above)
Comment (byronlin13) [reply .gitmodules:3]: Yup, this PR won't be merged until the object mapper is on NuGet and we remove the submodule.

File: Amazon.QLDB.Driver/client/Session.cs (hunk 22 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver/client/Session.cs]: Reverted the finally block to the base function

File: Amazon.QLDB.Driver/driver/QldbDriver.cs [hunk 28]
```diff
@@ -35,6 +35,7 @@ namespace Amazon.QLDB.Driver
     public class QldbDriver : IQldbDriver
     {
         private readonly QldbDriverBase<QldbSession> driverBase;
+        private readonly ISerializer serializer;
```
Comment (simonz-bq) [inline Amazon.QLDB.Driver/driver/QldbDriver.cs:38]: Looking at this in both drivers, can't this live as a part of the QldbDriverBase?

File: Amazon.QLDB.Driver/driver/QldbDriver.cs [hunk 29]
```diff
@@ -44,14 +45,17 @@ public class QldbDriver : IQldbDriver
...
             string ledgerName,
             IAmazonQLDBSession sessionClient,
             int maxConcurrentTransactions,
-            ILogger logger)
+            ILogger logger,
+            ISerializer serializer)
         {
             this.driverBase =
                 new QldbDriverBase<QldbSession>(ledgerName, sessionClient, maxConcurrentTransactions, logger);
+            this.serializer = serializer;
```
Comment (simonz-bq) [inline Amazon.QLDB.Driver/driver/QldbDriver.cs:58]: See above.

**Review by simonz-bq — CHANGES_REQUESTED** [review]

File: Amazon.QLDB.Driver/driver/QldbDriver.cs (hunk 28 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver/driver/QldbDriver.cs]: Good catch, moved `serializer` field to `QldbDriverBase`

File: Amazon.QLDB.Driver/driver/QldbDriver.cs (hunk 29 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver/driver/QldbDriver.cs]: Done

File: Amazon.QLDB.Driver/result/generic/AsyncResult.cs [hunk 30]
```diff
@@ -0,0 +1,173 @@
...
+        ///
+        /// <param name="session">The parent session that represents the communication channel to QLDB.</param>
+        /// <param name="txnId">The ID of the parent transaction.</param>
+        /// <param name="statementResult">The result of the statement execution.</param>
+        /// <param name="cancellationToken">The CancellationToken to use for this AsyncResult.</param>
+        /// <param name="query">An object encapsulating a query to be executed against QLDB.</param>
+        internal AsyncResult(
+            Session session,
+            string txnId,
+            QLDBSession.Model.ExecuteStatementResult statementResult,
```
Comment (simonz-bq) [inline Amazon.QLDB.Driver/result/generic/AsyncResult.cs:48]: Nit: Why not add QLDBSession.Model to the using list?

File: Amazon.QLDB.Driver/transaction/BaseTransaction.cs [hunk 31]
```diff
@@ -66,5 +67,16 @@ internal static QldbHash Dot(QldbHash seed, string statement, List<IIonValue> pa
 
             return seed.Dot(statementHash);
         }
+
+        internal static QldbHash Dot(QldbHash seed, string statement, ValueHolder[] parameters)
```
Comment (simonz-bq) [inline Amazon.QLDB.Driver/transaction/BaseTransaction.cs:71]: Why is this here/why do we need this?

**Review by simonz-bq — APPROVED** [review]

File: Amazon.QLDB.Driver/result/generic/AsyncResult.cs (hunk 30 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver/result/generic/AsyncResult.cs:48]: When I do that, 

`IOUsage` and `TimingInformation` on L77/87 are detected as 

* `Amazon.QLDBSession.Model.IOUsage`/
* `Amazon.QLDBSession.Model.TimingInformation`

instead of 
* `Amazon.QLDB.Driver.IOUsage`
* `Amazon.QLDB.Driver.TimingInformation`

So we would need to specify the complete namespace for iousage/timinginformation if we add QLDBSession.Model to the import list

File: Amazon.QLDB.Driver/transaction/BaseTransaction.cs (hunk 31 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver/transaction/BaseTransaction.cs:71]: It is to support the new [Transaction.execute()](https://github.com/awslabs/amazon-qldb-driver-dotnet/blob/object-serialization-idda-jun-4-2021/Amazon.QLDB.Driver/transaction/AsyncTransaction.cs#L148-L153) functions that take in the query object with `ValueHolder[]` as the parameters.

File: Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs (hunk 19 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs]: Updated to internal

File: Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs (hunk 20 above)
Comment (byronlin13) [reply Amazon.QLDB.Driver.IntegrationTests/utils/IntegrationTestBase.cs]: Updated to internal
